### PR TITLE
#3666 - Queue Monitoring - Default Queue Error Handling

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
This PR is a proposal to unify the queue behavior. It contains changes for start and cancel assessments regular queues and the idea is to expand it to all schedulers.

The below image shows these points to be shared and forced for all the queues, using the CAS scheduler as an example.
1 - Return type;
2 - Provide a `processSummary`;
3 - try/catch and default start message.
4 - Error check on `processSummary` to force the jobs to fail.
5 - Error handling on `catch`.
6 - Final log activities during `finally`.
image
_Note:_ The above will also ensure the E2E tests follow a closer approach.

Sample log for a queue using the new `BaseQueue<T>` class.

image

### Side effects
The idea is to have the BaseScheduler extend the BaseQueue which will force a change in every single scheduler (which would be the idea). To avoid the need to change every single scheduler and adapt every E2E, the methods required by BaseQueue can be implemented temporarily as below. The idea is to start the scheduler refactor still during this ticket.

```ts
/**
 * To be implemented.
 */
processQueue(_job: Job<void>): Promise<string | string[]> {
  throw new Error("Method not implemented.");
}

/**
 * To be implemented.
 */
async process(
  _job: Job<void>,
  _processSummary: ProcessSummary,
): Promise<string | string[]> {
  throw new Error("Method not implemented.");
}
```

### Error Fix
Fixed an issue with the logger service where `null` message was logged right after the regular message.

### CODE REVIEW

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-queue.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      const result = await this.process(job, processSummary);
+      processSummary.info(`${job.queue.name}, job ID ${job.id}, executed.`);
+      const logsSum = processSummary.getLogLevelSum();
+      if (logsSum.error) {
+        throw new CustomNamedError(
+          "One or more errors were reported during the process, please see logs for details.",
+          PROCESS_SUMMARY_CONTAINS_ERROR,
+        );
+      }
+      return result;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-queue.ts:38]: What about adding the attention message (`getSuccessMessageWithAttentionCheck`) when there is a warning like records ignored etc..

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-queue.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+   */
+  @Process()
+  async processQueue(job: Job<T>): Promise<string | string[]> {
+    const processSummary = new ProcessSummary();
+    try {
+      this.logger.log(`Processing queue ${job.queue.name}, job ID ${job.id}.`);
+      const result = await this.process(job, processSummary);
+      processSummary.info(`${job.queue.name}, job ID ${job.id}, executed.`);
+      const logsSum = processSummary.getLogLevelSum();
+      if (logsSum.error) {
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-queue.ts:32]: With logSum errors now certainly resulting in retry(being thrown) we should have a clear understanding with the team about when to use warning and error messages.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-queue.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  PROCESS_SUMMARY_CONTAINS_ERROR,
+  logProcessSummaryToJobLogger,
+} from "../../utilities";
+import { Job } from "bull";
+import { Process } from "@nestjs/bull";
+
+/**
+ * Provides basic functionality for queue processing.
+ */
+export abstract class BaseQueue<T> {
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-queue.ts:17]: It's good to have it in a abstract class. We can even have the logging for the job initialization and job ending in abstract methods if we that level of customization for some process. Good idea!

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-queue.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+   * Wrap the queue job execution in a try/catch for the global error handling.
+   * It provides also the default {@link ProcessSummary} for logging.
+   * @param job process job.
+   * @returns processing result.
+   */
+  @Process()
+  async processQueue(job: Job<T>): Promise<string | string[]> {
+    const processSummary = new ProcessSummary();
+    try {
+      this.logger.log(`Processing queue ${job.queue.name}, job ID ${job.id}.`);
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-queue.ts:28]: Great initiative to centralize the common logic. ❤️

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/start-application-assessment.processor.ts [hunk 5]
```diff
@@ -9,70 +9,60 @@ import {
...
+      `Starting assessment id ${job.data.assessmentId} using workflow ${workflowName}.`,
+    );
+    await this.workflowClientService.startApplicationAssessment(
+      workflowName,
+      job.data.assessmentId,
+    );
+    return "Workflow call executed with success.";
   }
 
   @InjectLogger()
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/assessment/start-application-assessment.processor.ts:68]: I believe the logger is not required here anymore.

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/start-application-assessment.processor.ts:68]: The logger should be added to every processor to ensure the right context is set (it is done during the property injection). Not having it causes the logs to be set to the base class context. We have many schedulers right now with this issue.

File: sources/packages/backend/apps/queue-consumers/src/processors/assessment/cancel-application-assessment.processor.ts [hunk 6]
```diff
@@ -169,17 +172,16 @@ export class CancelApplicationAssessmentProcessor {
...
+          processSummary.info(
             "No impacts were detected on future applications.",
           );
         }
       }
-      await summary.info("Assessment cancelled with success.");
-      return summary.getSummary();
+      return "Assessment cancelled with success.";
     });
   }
```
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/assessment/cancel-application-assessment.processor.ts:187]: Please check https://github.com/bcgov/SIMS/pull/4020#discussion_r1868424775

**Review by andrepestana-aot — APPROVED** [review]
Looks good to me. It will be very beneficial to the project and can be done in parts so I think we can at least start it in this sprint.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-queue.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-queue.ts:38]: It was supposed to 😉 
I am adding it now.

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/base-queue.ts]: Just added it.

**Review by dheepak-aot — APPROVED** [review]
Great work and thanks for brining this organization of things that are done in Queue processors. Looks good 👍

**Pull request merged**